File: clahe_improv.py
import cv2
import numpy as np

def main():
    cap = cv2.VideoCapture(0)

    # Set camera resolution
    cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)

    if not cap.isOpened():
        print("Cannot open camera")
        return

    print("Press 'q' to quit.")

    # Initialize previous edge frame for temporal smoothing
    prev_edges = None
    alpha = 0.6  # Weight for current frame in temporal smoothing

    # CLAHE setup
    clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))

    while True:
        ret, frame = cap.read()
        if not ret:
            print("Can't receive frame (stream end?). Exiting ...")
            break

        # Convert to grayscale
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

        # Apply CLAHE for contrast enhancement
        equalized = clahe.apply(gray)

        # Gaussian blur to reduce noise
        blurred = cv2.GaussianBlur(equalized, (5, 5), 1.5)

        # Canny edge detection with tuned thresholds
        low_thresh = 50
        high_thresh = 150
        edges = cv2.Canny(blurred, low_thresh, high_thresh)

        # Morphological closing to connect broken edges
        kernel = np.ones((3, 3), np.uint8)
        edges = cv2.morphologyEx(edges, cv2.MORPH_CLOSE, kernel)

        # Temporal smoothing with previous frame
        if prev_edges is not None:
            edges = cv2.addWeighted(edges, alpha, prev_edges, 1 - alpha, 0)
        prev_edges = edges.copy()

        # Show output
        cv2.imshow('Stable Real-Time Edge Detection', edges)

        # Quit on 'q' press
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cap.release()
    cv2.destroyAllWindows()

if __name__ == '__main__':
    main()

File: C++ code/Experiments/clahe_improv.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

int main() {
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Cannot open webcam!" << std::endl;
        return -1;
    }

    cap.set(cv::CAP_PROP_FRAME_WIDTH, 640);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, 480);

    cv::Mat frame, gray, clahe_gray, blurred, edges, previousEdges;
    cv::Ptr<cv::CLAHE> clahe = cv::createCLAHE(2.0, cv::Size(8, 8));

    std::cout << "Press 'q' to quit.\n";

    while (true) {
        cap >> frame;
        if (frame.empty()) break;

        // Preprocessing
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        clahe->apply(gray, clahe_gray);
        cv::GaussianBlur(clahe_gray, blurred, cv::Size(5, 5), 1.0);

        // Auto Canny using median
        double median = cv::mean(blurred)[0];
        double lower = std::max(0.0, 0.66 * median);
        double upper = std::min(255.0, 1.33 * median);
        cv::Canny(blurred, edges, lower, upper);

        // Optional: Smooth transition between frames (reduce flicker)
        if (!previousEdges.empty()) {
            cv::addWeighted(edges, 0.7, previousEdges, 0.3, 0, edges);
        }
        previousEdges = edges.clone();

        // Thin dilation to preserve structure
        cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(2, 2));
        cv::dilate(edges, edges, kernel);

        cv::imshow("Stable Canny Edge Detection", edges);

        char key = (char)cv::waitKey(1);
        if (key == 'q') break;
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}
